### ISSUE TITLE
Execution time appears for a short time, then disappears

### ISSUE THREAD

**Issue by infokiller** [romkatv/powerlevel10k#171]
Hi Roman,

I just updated powerlevel10k to the latest version (e6d0da1c2e1b95e2a0169b7abcd6ef8b7650b4bf) and used `p10k config` to generate a configuration which is similar to the old `purepower` (which I now understand is replaced by the lean style?).
One thing I noticed is that I no longer see the execution time of the last command in the right prompt **only in a certain git repo**. After playing with it a bit, I noticed that it does actually show up, but then disappears after a few milliseconds. I used the following settings for testing this:
```zsh
typeset -g POWERLEVEL9K_COMMAND_EXECUTION_TIME_THRESHOLD=0
typeset -g POWERLEVEL9K_COMMAND_EXECUTION_TIME_FOREGROUND=1
```

I can upload a screencast later to show the problem and will invest more time in debugging this, but wanted to know if you have any idea what could cause this?

Thanks!

**Comment by romkatv** [comment]
Thanks for the bug report! Fixed.

> I just updated powerlevel10k to the latest version ([e6d0da1](https://github.com/romkatv/powerlevel10k/commit/e6d0da1c2e1b95e2a0169b7abcd6ef8b7650b4bf)) and used `p10k config` to generate a configuration which is similar to the old `purepower` (which I now understand is replaced by the lean style?).

Yes, lean style is based on Pure Power.

By the way, you don't have to update your configs if you don't want to. All configs are supported forever and your prompt will stay just the way it is even if you update powerlevel10k. It might get faster but it'll look the same as before. The only reason to update your config is if you want your prompt to look different. Granted, the new configs are really good, so I do recommend switching over, but there is no pressure and never will be. It's also the reason why `p10k configure` dumps a config in your home directory, which you can add to your dotfiles repo if you have it. This allows me to change config templates in p10k so that new users will pick up new features but existing users won't be annoyed by constant prompt UI churn.

**Comment by infokiller** [comment]
Thanks for fixing this so quickly, always a pleasure consuming your software!

Re the config update - my motivation for switching was that I wanted to be able to regularly pull the improvements you're making to the config. I currently manage powerlevel10k using a git submodule in my dotfiles repo, and I intend to periodically sync the submodule to the latest master commit, and possibly rerun `p10k config` to get the latest and greatest config, which I indeed manage in my dotfiles repo as well.
I do have some minor customizations in my `zshrc` after I'm sourcing the output of `p10k config`. For example right now this is what I have in my `zshrc`:
```zsh
source "${ZSH_CONFIG_DIR}/p10k.zsh"
# When running as root, show the context (user@hostname) in red to make it hard
# to miss.
declare -g POWERLEVEL9K_CONTEXT_ROOT_FOREGROUND=196
# NOTE: As of 2019-08-16, the conda environment is already showed on the right
# prompt by default.
# POWERLEVEL9K_LEFT_PROMPT_ELEMENTS+=(custom_lprompt)
# declare -g POWERLEVEL9K_CUSTOM_LPROMPT=custom_lprompt
# declare -g POWERLEVEL9K_CUSTOM_LPROMPT_BACKGROUND=none
# declare -g POWERLEVEL9K_CUSTOM_LPROMPT_FOREGROUND=none
# custom_lprompt() { [[ -n "${CONDA_DEFAULT_ENV}" ]] && printf '%s' "(${CONDA_DEFAULT_ENV})" }
POWERLEVEL9K_RIGHT_PROMPT_ELEMENTS+=(custom_rprompt)
declare -g POWERLEVEL9K_CUSTOM_RPROMPT=custom_rprompt
declare -g POWERLEVEL9K_CUSTOM_RPROMPT_BACKGROUND=none
declare -g POWERLEVEL9K_CUSTOM_RPROMPT_FOREGROUND=none
# Mark prompt with ranger if it was opened from it. Copied from:
# https://github.com/ranger/ranger/blob/master/examples/bash_subshell_notice.sh
custom_rprompt() { [[ -n "${RANGER_LEVEL}" ]] && printf '%s' '(ranger)' }
# NOTE: As of 2019-08-16, this is no longer needed.
# DEFAULT_USER must be set so that user@hostname is not printed for the current
# user.
# if [[ ${EUID} -ne 0 ]]; then
#   declare -g DEFAULT_USER="${USER}"
# fi
source "${PLUGINS_DIR}/powerlevel10k/powerlevel10k.zsh-theme"
```
I should also note that I don't put my customizations directly in `p10k.zsh` (the output from `p10k config`) because I wanted an easy way to update that file without overriding my changes.
Is this a reasonable way to customize powerlevel10k?

BTW, one limitation I hit with this approach is that I couldn't customize the git branch color. It used to be grey for me in `purepower`, but when I switched to the output from `p10k config` it became green. I was able to change it back by editing `p10k.zsh`, but couldn't find a variable that controls this so that I can do this change only after I source `p10k.zsh`. For reference, in my generated `p10k.zsh` I changed this line to customize the branch color:
```zsh
  # If on a branch...
  vcs+='${${VCS_STATUS_LOCAL_BRANCH:+%76F'${(g::)POWERLEVEL9K_VCS_BRANCH_ICON}
```

**Comment by romkatv** [comment]
> `custom_rprompt() { [[ -n "${RANGER_LEVEL}" ]] && printf '%s' '(ranger)' }`

There is a better and faster API for custom prompt segments. Search for `prompt_example` in `~/.p10k.zsh`.

Specifically for ranger, I just added `ranger` prompt, so you don't have to use a custom segment for it. It's enabled by default in configs generated by `p10k configure`, although it's on the first line rather than second as you had it before. You can move it of course.

> Is this a reasonable way to customize powerlevel10k?

This approach assumes that setting `POWERLEVEL9K_CONTEXT_ROOT_FOREGROUND` after sourcing `~/.p10k.zsh` will do the same thing in the future after you generate a new version of `~/.p10k.zsh`. This is not guaranteed.

For a less fragile approach, edit `~/.p10k.zsh` directly and use git to merge your changes with upstream. First, copy [p10k-lean.zsh](https://github.com/romkatv/powerlevel10k/blob/master/config/p10k-lean.zsh) or [p10k-classic.zsh](https://github.com/romkatv/powerlevel10k/blob/master/config/p10k-classic.zsh) to your dotfiles repo under the name of `~/.p10k.zsh`. Commit. Run `p10k configure` and let it overwrite `~/.p10k.zsh`. Commit. Finally, apply your manual edits to `~/.p10k.zsh` and commit those as well. You now have two commits on top of the stock config.

To rebase your modifications on top of a newer version of `p10k-lean.zsh`/`p10k-classic.zsh` do the following. Copy [p10k-lean.zsh](https://github.com/romkatv/powerlevel10k/blob/master/config/p10k-lean.zsh) or [p10k-classic.zsh](https://github.com/romkatv/powerlevel10k/blob/master/config/p10k-classic.zsh) to your dotfiles repo under the name of `~/.p10k.zsh`. Commit. Cherry pick your two edit commits one by one, resolve conflicts if any, and commit.

There are many other ways how you can utilize git to rebase your modifications on top of the newer base config. None of them are easy (you'll have to resolve conflicts for sure) because it's not a directly supported use case. `p10k-lean.zsh`/`p10k-classic.zsh` are made specifically to give me freedom to change things without any backward-compatibility guarantees. It's how I can give decent prompt to new users without breaking old users. Powerlevel9k doesn't have this layer, so 99% of the userbase is stuck with a borderline unusable prompt as they are incapable of using anything other than the default configuration.

**Comment by infokiller** [comment]
Thanks so much! I'll follow your recommendations.

**Comment by infokiller** [comment]
> 
> > Is this a reasonable way to customize powerlevel10k?
> 
> This approach assumes that setting `POWERLEVEL9K_CONTEXT_ROOT_FOREGROUND` after sourcing `~/.p10k.zsh` will do the same thing in the future after you generate a new version of `~/.p10k.zsh`. This is not guaranteed.

Reading this again, I'm not sure I understand this. Why is setting `POWERLEVEL9K_CONTEXT_ROOT_FOREGROUND` after sourcing `~/.p10k.zsh` more fragile than editing `~/.p10k.zsh` directly?
I mean, if I were to add this change to the file directly, I would probably just add very similar lines below the original definition of that setting.
> 
> For a less fragile approach, edit `~/.p10k.zsh` directly and use git to merge your changes with upstream.

**Comment by romkatv** [comment]
In the future version of `p10k-lean.zsh` there may be no context. Or its colors might be configured via `CONTENT_EXPANSION` so that `FOREGROUND` definitions will have no effect (kinda like it is right now with `vcs`). Or context may be hidden when you are root and perhaps some other segment will be shown instead. If any of these happen, your prompt will silently break if you simply define parameters after sourcing `p10k-lean.zsh`. But if you instead locally change this line in your config: https://github.com/romkatv/powerlevel10k/blob/49063ee757c2339c2a0cde14e886896373cd6a54/config/p10k-classic.zsh#L437
then you'll have a merge conflict (it's a good thing in this case) and will have to figure out what to do.

Makes sense?

At the end of the day, it's up to you to decide which workflow has the right trade off between work and risk. As long as you understand that there are no guarantees how `p10k-lean.zsh` interacts with extra options that are set after it, you can make the right choice for yourself.

**Comment by romkatv** [comment]
Just to reiterate, the configs are optimized for the sole purpose of serving new users. If I figure out a way to make configs a little bit better for new users -- either by changing the visible properties of the prompt or restructuring the config so that it's easier to modify -- I'll do that even if it means that "upgrading a config" for the existing users will be hard. This is because there is no real need to upgrade a config. If you want to follow along with the evolution of the new-user-config, you can do that but you are on your own.

I'll also note that your config changes are really small. If I understand it correctly, it's just the color of root context (yellow => red) and the color of local git branch (green => grey). The latter is probably not a good idea because it'll be difficult or impossible to distinguish stale prompts from up-to-date prompts. The former may not be important in practice. Do you ever open zsh as root? with all your plugins and themes? I would advice against this. If you want your prompt to pick up new-user-config changes, you can do this much easier if you forgo your (small) customizations. Just another option to consider.

**Comment by infokiller** [comment]
Thanks for the explanation! Makes sense. It's basically a tradeoff between having a risk of an unexpected change, or the risk of a merge conflict (which could be just a false positive) with a lower chance of an unexpected change.

> If I understand it correctly, it's just the color of root context (yellow => red) and the color of local git branch (green => grey). The latter is probably not a good idea because it'll be difficult or impossible to distinguish stale prompts from up-to-date prompts. The former may not be important in practice. Do you ever open zsh as root? with all your plugins and themes? I would advice against this.

Re the git branch color- is there a way to distinguish between stale and up-to-date prompts by default that is being suppressed by this change?

Re running zsh as root with custom settings- I don't do that often, and I agree it's probably a bad idea because the consequences of misconfiguration can be greatly amplified. I don't even recall why I wanted to do it...

**Comment by romkatv** [comment]
> Thanks for the explanation! Makes sense. It's basically a tradeoff between having a risk of an unexpected change, or the risk of a merge conflict (which could be just a false positive) with a lower chance of an unexpected change.

That's right.

> Re the git branch color- is there a way to distinguish between stale and up-to-date prompts by default that is being suppressed by this change?

By default stale Git prompt is grey. You can only see it in massive Git repositories. If you make non-stale Git branch name grey, it'll be indistinguishable from stale. It may not be that big of a deal but it's still information loss.
